### DISCUSSION TITLE
WixToolset.DifxApp removed in v5 - alternatives?

### DISCUSSION THREAD

**Discussion by x86dev** [wixtoolset/issues#8153] [Questions]
As WixToolset.DifxApp got deprecated in v4 and removed in v5, what are the recommended alternatives to install drivers now? Is v5 getting an adequate replacement for WixToolset.DifxApp so that users don't have to invent custom (action) solutions for each and every installer which is using v5?

**Comment by x86dev** [comment]
Any news here?

**Reply by robmen** [reply]
> so that users don't have to invent custom (action) solutions for each and every installer which is using v5

Microsoft deprecated DifxApp, so why don't you contribute its replacement to the WiX Toolset?

**Reply by x86dev** [reply]
Before intending so, what is the actual take by the WIX team regarding this? Is there anything in the make already or at least something planned? An official statement besides "Microsoft has deprecated DifxApp" would help here.

**Reply by robmen** [reply]
I'm not aware of anyone starting to work on a replacement for DifxApp.

**Comment by computerquip-work** [comment]
Looking at MSDN, I'm not even sure they want drivers to be distributed in an MSI anymore. They're pretty clear in saying that they want drivers to be packaged "individually" and to submit it for distribution through Windows Update channels.

This is a very "corporate" solution to me and not even a practical one. Like imagine if nvidia only distributed their drivers through Windows update. We wouldn't see updates but twice a year at best because of all the non-sense that has to go into getting a driver updated through official channels.

As a footnote on MSDN, they say that applications using DIFx should consider switching to using pnputil.exe or [DiInstallDriver](https://learn.microsoft.com/en-us/windows/win32/api/newdev/nf-newdev-diinstalldriverw) and [DiUninstallDriver](https://learn.microsoft.com/en-us/windows/win32/api/newdev/nf-newdev-diuninstalldriverw).

So as far as I see it, ideally there would be some WIX extension implemented over those two functions and in the meantime, use a CustomAction with pnputil.exe.

EDIT: Something that totally escaped me during the thought process above is that pnputil.exe has no convenient way to uninstall a driver. You generally specify a `.inf` file of some sort which gets changed to `oem<n>.inf` where you don't know what `<n>` is. There's no way to know how to uninstall it ahead of time so it's very difficult to write declarative rules concerning driver uninstall unfortunately.

**Reply by wmanning** [reply]
There's also the DevicePath key in the registry: https://learn.microsoft.com/en-us/windows-hardware/drivers/install/how-windows-selects-a-driver-for-a-device

**Reply by computerquip-work** [reply]
Also, FireGiant does seem to support a driver installation extension as part of their HeatWave Build Tools solution: https://www.firegiant.com/products/heatwave-build-tools/#heatwave-harvesting

Seems to require paid enterprise support.

**Reply by robmen** [reply]
Note: the FireGiant Driver Extension is for non-PnP drivers (think file system drivers). Read more on the documentation page: https://www.firegiant.com/docs/heatwave-build-tools/driver/.

It is not a replacement for DifxApp, which Microsoft created to install PnP drivers.

**Reply by computerquip-work** [reply]
I didn't notice that, thanks for the clarification.

**Reply by computerquip-work** [reply]
I edited my original post to include some problems with pnputil.exe during uninstall. If uninstall of drivers is a requirement, I actually don't think there's a way to do so currently with Wix v5 without support tools or external help.

**Reply by computerquip-work** [reply]
So... I tried writing a dirty solution to this. I failed pretty miserably and don't have any more time to work on this. I'm going to leave some notes here.

First off, OEM INF files are created using the process described [here](https://learn.microsoft.com/en-us/windows/win32/api/setupapi/nf-setupapi-setupcopyoeminfa). However, that does *not* describe how it gets installed into the driver store. There's a particular piece there that's kind of important: `When SetupCopyOEMInf copies a file, it renames the copied file to OEM*.inf. Name provided is unique and cannot be predicted.` DiInstallDriver does not provide the OEM file location or name thus it cannot be used i.e. this function is useless outside of simple use cases. It also has various restrictions on what kind of drivers that it may install. 

Seemingly, DefaultUninstall is an unreliable field of the INF and cannot be used reliably.

SetupCopyOEMInf is deprecated, along with all of the other general setup functions. Even if you used it, it doesn't solve 90% of the rest of the problems the solution requires.

The only way I see it to be possible is to query every `oem*.inf` file in `%SystemRoot%\INF` or to find similar named .inf files in the undocumented Driver Store folder, do a checksum and compare, then call DiUninstallDriver on the INF if matched. There's no guarantee it's the same driver that was installed or not really but I don't believe DIFx did much better. There's still a whole slew of issues there unfortunately.

I pretty much gave up after that realization. I have no idea what this driver model is even going for. Coming from the Linux world, where driver handling can be as simple as one command with 2 arguments, I cannot comprehend the ridiculous amounts of complexity for such an important part of the OS ecosystem. My only guess is they literally want to prevent you from uninstalling a driver which doesn't make any sense for things like FS filters or decrepit drivers with vulnerabilities. Or maybe they seriously believe you should only distribute through them? That is completely outside of reality if they actually think that.

Anyways, frustrating problem and I have to revert back to DIFx and WIX4 for now.

**Reply by wmanning** [reply]
For driver installation I guess you can do:
1. Use MSI to install driver files (including inf file) to your chosen location (say C:/Program Files/myapp/mydriver)
2. Use DiInstallDriver.  Pass in the location of the INF file in step one

It appears uninstallation is more complicated like you ran into.

According to this https://learn.microsoft.com/en-us/windows-hardware/drivers/install/how-devices-and-driver-packages-are-uninstalled#deleting-a-driver-package-from-the-driver-store
- (On Windows 10, version 1703 and latter) you can use DiUninstallDriverW.  Wonder if you can pass in the location of the inf file installed by the MSI in 1. above or if you need the OEM name in the windows INF folder?

**Reply by computerquip-work** [reply]
I've tried installing and uninstalling a custom driver from a specific location to see what it would do.

`DiInstallDriver` with `C:\MyDriver\MyDriver.inf` -> `ERROR_SUCCESS`
I can see the driver listed by `pnputil.exe /enum-drivers`, it has an `oem*.inf` file and shows the original name.
Immediately follow that with `DiUninstallDriver` with `C:\MyDriver\MyDriver.inf` -> `ERROR_NOT_FOUND`. The documentation doesn't explicitly explain what INF you can pass to DiUninstallDriver but I'm guessing it's based on the Remarks found [here](https://learn.microsoft.com/en-us/windows/win32/api/setupapi/nf-setupapi-setupgetinfdriverstorelocationw).

**Reply by wmanning** [reply]
Too bad. Yeah it's annoying the documentation is so unclear.

**Reply by wmanning** [reply]
To add if you install the drivers using DiInstallDriver and go into Add/remove programs (called Apps & features now) do you see your driver? Should start with "Windows Driver Package - ...".

I'm wondering if a add/remove entry gets created for all drivers installed to the driver store. If so at least there's a way for users to uninstall the driver.

**Reply by computerquip-work** [reply]
Not that I see. From a user-perspective, the only way I can tell to uninstall the driver at that point is `pnputil.exe /delete-driver` after finding the OEM file via `pnputil.exe /enum-drivers`.

**Reply by bmtpolo** [reply]
FYI, if you copy your driver to the filesystem, you don't need to know how pnputil assigns it's oem number.  For instance,
cd "Program Files"/DummyApplication
pnputil /add-driver dummy_driver.inf /install

For me it got assigned as oem8.inf

Then on uninstall 
cd "Program Files"/DummyApplication
pnputil /delete-driver dummy_driver.inf

And it's successful for me (At least from the command line).  I'm still trying to get the install to work via custom action and quiet exec

**Reply by computerquip-work** [reply]
I have notes that say this didn't work but that was a couple years ago now. Maybe I did something weird then... It seems to work on my VMs at least now.

**Reply by wmanning** [reply]
This [link](https://learn.microsoft.com/en-us/windows-hardware/drivers/install/how-devices-and-driver-packages-are-uninstalled#deleting-a-driver-package-from-the-driver-store) says it was introduced in Windows 10 1607.
>Deleting a Driver Package from the Driver Store
>...
>- Starting in Windows 10, version 1607, from the command prompt, use pnputil /delete-driver <example.inf> /uninstall. For info on PnPUtil commands, see [PnPUtil Command Syntax](https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/pnputil-command-syntax).
>- Starting in Windows 10, version 1703, a device installation application can call [DiUninstallDriverW](https://learn.microsoft.com/en-us/windows/win32/api/newdev/nf-newdev-diuninstalldriverw).

**Reply by computerquip-work** [reply]
Yeah, `delete-driver` existed back then but I'm fairly sure you could only pass `oem#.inf` as described in the help and documentation:
![image](https://github.com/wixtoolset/issues/assets/69005729/ea86ffd0-e7b5-4178-9476-3264908a5cb6)

**Reply by computerquip-work** [reply]
Yeah, okay, I'm not crazy, this doesn't work on Windows Server 2019 and maybe early versions of Windows Server 2022.

**Reply by robmen** [reply]
The next time someone asks, "Why haven't you provided a replacement for DifxApp?" please point them to this thread.

It's a non-trivial problem and very disappointing that Microsoft deprecated DifxApp.

**Comment by computerquip-work** [comment]
With wixv3 and v4 apparently having support dropped, I'm once again coming back to this god forsaken problem. If I were to hypothetically start work on something for this...

Is a generic solution possible here at all? What scenarios would I need to target to make everyone happy?

**Reply by robmen** [reply]
Start by creating the C functions to detect, install and uninstall PnP drivers. Take a look at the existing WiX custom action C code to see the style and how dutil.lib helps standardize the code.

Once you have those functions written and we know what inputs they take to do their work, we can build the MSI integration on top.

Those first three functions (detect, install, uninstall) require the specific driver domain knowledge that us setup developers don't have.
